### PR TITLE
Use writeShellApplication to validate shell scripts

### PR DESCRIPTION
**Pull request by jfroche**
`writeShellApplication` validates shell scripts included in Nix with shellcheck and `bash -n`

### CODE REVIEW

**Review by soenkehahn — CHANGES_REQUESTED** [review]
Could you describe a bit what this PR does and why? It looks like it switches some test scripts over from `writeScript` to `writeShellScriptBin`. Is that true? Does it change anything else? Also, why is that a good idea? The title says '... to validate ...'. Does `writeShellScriptBin` perform some kind of validation? I can't find documentation about that. Also, why is `writeShellScriptBin` better than `writeShellScript`?

**Comment by r-vdp** [comment]
`writeShellScript{.Bin}` run the script in dry-run mode during the build to validate its syntax, meaning that you get a build-time error instead of a run-time error in case of syntax errors, which means that syntax errors get caught in CI instead of in production.

There's also `writeShellApplication` which additionally runs `shellcheck` on the scripts, doing even more validation. I usually prefer this where ever possible.

**Comment by soenkehahn** [comment]
> writeShellScript{.Bin} run the script in dry-run mode during the build to validate its syntax

What's dry-run mode? Also, where is this behavior of `writeShellScriptBin` documented, if anywhere?

**Comment by r-vdp** [comment]
> > writeShellScript{.Bin} run the script in dry-run mode during the build to validate its syntax
> 
> What's dry-run mode? Also, where is this behavior of `writeShellScriptBin` documented, if anywhere?

Mainly in the source code, I'm afraid: https://github.com/NixOS/nixpkgs/blob/3d19c43d6045093cc29ab251ad9919fe0296e6c5/pkgs/build-support/trivial-builders/default.nix#L280

The dry-run mode is a feature of bash where it basically parses the script and checks syntax, but doesn't actually run any commands.

And for `writeShellApplication`: https://github.com/NixOS/nixpkgs/blob/3d19c43d6045093cc29ab251ad9919fe0296e6c5/pkgs/build-support/trivial-builders/default.nix#L311

**Comment by jfroche** [comment]
We can also use `writeShellApplication` to use shellcheck to also validate these scripts

File: flake.nix [hunk 1]
```diff
@@ -83,10 +83,16 @@
...
+              let
+                testFileServerScript = pkgs.writeShellApplication {
+                  name = "test-file-server";
+                  text = ''
+                    cd ${packages.exampleInstallerFiles}
+                    exec ${pkgs.python3}/bin/python3 -m http.server 2> /dev/null
+                  '';
+                };
+              in
+              "${testFileServerScript}/bin/test-file-server";
```
Comment (soenkehahn) [inline flake.nix]: Can we use `pkgs.lib.getExe` and inline `testFileServerScript`?

File: flake.nix [hunk 2]
```diff
@@ -118,43 +124,54 @@
...
+                      -device e1000,netdev=net0 \
+                      -netdev user,id=net0,hostfwd=tcp::2222-:22 \
+                      &
+                    # Wait for the runcmd to call us back to let us know that we are fully booted.
+                    # (See runcmd in the flake file under cloudcfg)
+                    nc -vl 2223
+                  '';
+                };
+              in
+              "${bootVMScript}/bin/boot-vm";
```
Comment (soenkehahn) [inline flake.nix]: Same here, maybe use `pkgs.lib.getExe` instead of string interpolation?

File: flake.nix [hunk 3]
```diff
@@ -118,43 +124,54 @@
...
+                      -o "StrictHostKeyChecking no" \
+                      -o "UserKnownHostsFile=/dev/null" \
+                      localhost \
+                      -p 2222 \
+                      -l test \
+                      "$@"
+                  '';
+                };
+              in
+              "${sshVMScript}/bin/ssh-vm";
```
Comment (soenkehahn) [inline flake.nix]: Same here.

**Review by soenkehahn — APPROVED** [review]
Apart from my comments LGTM.

**Pull request merged**